test/e2e/framework/rc_util.go

```diff
@@ -247,7 +247,7 @@ func ValidateController(c clientset.Interface, containerImage string, replicas i
 	// You can read about the syntax here: http://golang.org/pkg/text/template/.
 	getContainerStateTemplate := fmt.Sprintf(`--template={{if (exists . "status" "containerStatuses")}}{{range .status.containerStatuses}}{{if (and (eq .name "%s") (exists . "state" "running"))}}true{{end}}{{end}}{{end}}`, containername)
 
-	getImageTemplate := fmt.Sprintf(`--template={{if (exists . "status" "containerStatuses")}}{{range .status.containerStatuses}}{{if eq .name "%s"}}{{.image}}{{end}}{{end}}{{end}}`, containername)
+	getImageTemplate := fmt.Sprintf(`--template={{if (exists . "spec" "containers")}}{{range .spec.containers}}{{if eq .name "%s"}}{{.image}}{{end}}{{end}}{{end}}`, containername)
```

Comment:
This code is legacy, and it's a bit obtuse to follow why this change is needed for multi-arch.  I'm not following why it doesn't just work.

Reply:
No problem, I can give more context.

I run to this problem because one of our end-to-end test in ARM was failing, emitting this error:

```
Oct 26 11:05:31.784: INFO: update-demo-kitten-7gx94 is created but running wrong image; expected: gcr.io/kubernetes-e2e-test-images/kitten:1.0, actual: gcr.io/kubernetes-e2e-test-images/kitten-arm64:1.0
```

As you can see, the image expected does not include the '-arm64' arch suffix.

It seems that the test does no include it, since they are platform agnostic, so the image name is a generic one (the one without the '-arm64' suffix). But this image name is a multiarch one, so this name only points to a manifest, that include the real image name for all the archs that are actually supported. For example:

```
$ kubectl describe pod -n e2e-tests-kubectl-fdqr8 update-demo-kitten-zj424
Name:               update-demo-kitten-zj424
Namespace:          e2e-tests-kubectl-fdqr8
Priority:           0
PriorityClassName:  <none>
Node:               x-5/192.168.2.55
Start Time:         Fri, 26 Oct 2018 13:38:26 +0000
Labels:             name=update-demo
                   version=kitten
Annotations:        midonet.org/mac-address=42:88:73:f6:38:5e
Status:             Running
IP:                 10.1.1.136
Controlled By:      ReplicationController/update-demo-kitten
Containers:
 update-demo:
   Container ID:   docker://3890dad2629483535ff1e86e25dd4e36f33415626fc85c35ff331950969aad21
   Image:          gcr.io/kubernetes-e2e-test-images/kitten:1.0
   Image ID:       docker-pullable://gcr.io/kubernetes-e2e-test-images/kitten-arm64@sha256:fbc805315cd3fc3adfa90848959193ae45808444e1eb9af855d32f1dfe11ded6
```

You can see that the value of the 'Image' field is different than the 'Image ID', which includes the actual image name and its SHA256 hash.

The code in the test seems to be using not the image name in the specification (under '.spec') of the container, but the image in the status (under '.status'), as can be seen by this log of the command that retrieves it (in kubernetes v1.12.1):

```
Oct 26 13:43:01.540: INFO: Running '/usr/bin/kubectl --kubeconfig=/home/nvidia/jenkins/workspace/iessp/iaas/kubernetes-testing/midonet_kubernetes_e2e_arm64/46/iessp-iaas/tests/ci/k8s/clusters/sshlogin/kubeconfig get pods update-demo-kitten-hsgdw -o template --template={{if (exists . "status" "containerStatuses")}}{{range .status.containerStatuses}}{{if eq .name "update-demo"}}{{.image}}{{end}}{{end}}{{end}} --namespace=e2e-tests-kubectl-fdqr8'
Oct 26 13:43:01.931: INFO: stderr: ""
Oct 26 13:43:01.931: INFO: stdout: "gcr.io/kubernetes-e2e-test-images/kitten-arm64:1.0"
```

Since the name retrieved by the command includes the '-arm64' suffix, the test code seems it is wrong.

I don't know if nobody is using this code (kubernetes v1.12.1) on ARM64, or if this happening just in specific versions of the container runtime (specific Docker versions).

Is this code legacy now? no longer used in master? I didn't know, sorry. With this changes our tests pass correctly.

Anyway, I think the test code is more platform agnostic if the command retrieving the image name uses the 'spec' instead of the 'status', since the test code is no able to guess what is the name of the specific images per arch.

Thanks for your time. Let me know what you think and if anything else needs clarification.

Reply:
This makes perfect sense now. 
/lgtm 
/approve